### PR TITLE
feature: Enable Context

### PR DESCRIPTION
**Pull request by kjjuno**

### CODE REVIEW

**Comment by kjjuno** [comment]
This replaces https://github.com/David-Kunz/gen.nvim/pull/31 as that PR was accidentally closed during merge conflict resolution.

**Comment by David-Kunz** [comment]
Thanks, @kjjuno !

I think we should merge that, but I want to do some additional testing before.

**Comment by kjjuno** [comment]
What are your thoughts about the container? I can’t run curl against the container so I think we would have to rely on them opening a port to the container and just using the REST api.

**Comment by kjjuno** [comment]
> I get the error
> 
> ```
> ^I[C]: in function 'nvim_buf_get_lines'
> ^I/Users/d065023/projects/nvim/gen.nvim/lua/gen/init.lua:32: in function 'write_to_buffer'
> ^I/Users/d065023/projects/nvim/gen.nvim/lua/gen/init.lua:245: in function </Users/d065023/projects/nvim/gen.nvim/lua/gen/init.lua:216>
> ```
> 
> when closing the split and running `:Gen` with `Change Code`.

I’m guessing this is similar to the last bug you reported. I’ll see if I can fix it later today.

**Comment by kjjuno** [comment]
The invalid buffer bug should be fixed now.

**Comment by kjjuno** [comment]
> I also prefer not to show the prompt, especially for longer prompts, this is quite distracting. Users usually are only interested in the result.
> 
> In the future:
> 
> * we might also allow follow up questions
> * we might provide actions to replace the previously highlighted text on demand (and remove the `replace = true` option)

I personally like the prompt history, especially in a longer iterative chat. However, your point on the longer context makes a lot of sense. How would you feel about showing the first 2 lines or something? I could even wrap that in a feature flag so that you can opt into displaying the prompt and having the default behavior be to hide it.

I'll play around with that idea and push something up.

I also have a local version of https://github.com/David-Kunz/gen.nvim/pull/22 that uses curl instead of plenary. I enjoy using it. I would be happy to add it to this PR if you would be interested. If not I'll just use that locally until https://github.com/David-Kunz/gen.nvim/pull/22 get's merged.

**Comment by kjjuno** [comment]
I'm thinking of making it look something like this

```lua
require("gen").setup({
  model = "codellama:34b",
  show_prompt = true,
})

vim.keymap.set({ "n", "v" }, "<leader>go", ":Gen<CR>")
vim.keymap.set({ "n", "v" }, "<leader>gc", ":Gen Chat<CR>")

```

**Comment by kjjuno** [comment]
image

**Comment by David-Kunz** [comment]
Thank you, @kjjuno . I still find it too much visual clutter, I would also remove the headline `# Chat with <model>`, that's something the user already knows. Hence I propose to remove everything except the generated text, but since you like to add the prompt, maybe we can add it optionally?

```lua
require('gen.nvim').show_prompt = true
```

And I would show the separator only if a new result is appended.

Otherwise, the PR is good and seems to work on my machine 👍 

What do you think?

Thank you and best regards,
David

**Comment by leafo** [comment]
Just wanted to add a note as a daily user of this plugin: I think the change to a split is subjective and would be something I personally would not like. I code with splits, and I feel like plugins that insert a new split often do it under the assumption that someone is just looking at a single file. When you already have splits on the screen and a plugin inserts a new split the shift in layout can be very jarring and annoying. I think a floating window is perfect fit for this library, as the result of the prompt is a temporary action you're making on the code you're editing.

**Comment by kjjuno** [comment]
> Just wanted to add a note as a daily user of this plugin: I think the change to a split is subjective and would be something I personally would not like. I code with splits, and I feel like plugins that insert a new split often do it under the assumption that someone is looking at a single file. When you already have splits on the screen and a plugin inserts a new split the shift in layout can be jarring and annoying. (I'm trying to ask the LLM a question, not re-arrange my editor!) I think a floating window is perfect fit for this library, as the result of the prompt is a temporary action you're making on the code you're editing.
> 
> Edit: Also regarding showing the prompt, I agree it's visual clutter. I tend to have very big prompts for most of my queries as I'm sending in the majority of the text on my screen with my average prompt. If displaying the prompt or part of it is added, then I vote for having the ability to turn off displaying it in the output window.

I would be happy to make the change to splits an opt in feature to maintain what current users are used to.

**Comment by kjjuno** [comment]
@leafo I should also note that the displaying of the prompt is off by default and is something you would have to opt into.

**Comment by kjjuno** [comment]
That’s been pushed

**Comment by kjjuno** [comment]
I still think the docker container feature needs to be considered before merging. As of now this PR breaks the direct integration with a container and would require manual maintenance of your docker container and adds the requirement that they bind a port.

**Comment by kjjuno** [comment]
unfortunately that last push introduced some new edge case bugs when the buffer window is closed, or the buffer is deleted. I'm still debugging them.

**Comment by David-Kunz** [comment]
Thank you @leafo for your comments and thank you @kjjuno for taking those into account, I tend to agree.

Pro vertical split:
- can be left open (see https://github.com/David-Kunz/gen.nvim/issues/39)

Pro float:
- more horizontal space

I tend to prefer floats.

@kjjuno I tried your branch, but I don't get any responses. We should also enable

```lua
    pcall(io.popen, "ollama serve > /dev/null 2>&1 &")
```

if no container is used.

We previously had the `M.container` option, was there a reason you removed it? I think we should restore the former container behaviour (allow to set `M.container` and change the default command to run it).

What do you think?

**Comment by David-Kunz** [comment]
> @kjjuno I tried your branch, but I don't get any responses. We should also enable

@kjjuno now it seems to work, not sure what happened...

File: lua/gen/init.lua [hunk 1]
```diff
@@ -41,54 +67,111 @@ local function get_window_options()
...
+            end
+            reset()
+        end,
+    })
+end
+
+function reset()
+    M.result_buffer = nil
+    M.float_win = nil
+    M.result_string = ""
```
Comment (David-Kunz) [inline lua/gen/init.lua:133]: Can we put `result_string` into the closure? It doesn't need to be part of the module, no?

File: lua/gen/init.lua [hunk 2]
```diff
@@ -113,98 +196,203 @@ M.exec = function(options)
...
+        prompt = prompt,
+        stream = true,
+    }
+
+    if M.context then
+        bodyData.context = M.context
+    end
+
+    local json = vim.fn.json_encode(bodyData)
+    local cmd = "curl --silent --no-buffer -X POST "
```
Comment (David-Kunz) [inline lua/gen/init.lua:219]: Can we also support `M.command`?

Comment (David-Kunz) [reply lua/gen/init.lua:219]: (otherwise it would be a breaking change)

Comment (kjjuno) [reply lua/gen/init.lua:219]: The biggest problem there is that the container that hosts ollama doesn’t contain curl. I haven’t yet figured out a good way to handle that. If you have suggestions I would love ideas

Comment (David-Kunz) [reply lua/gen/init.lua:219]: I see, but we could at least allow users to plug their own commands in there, maybe they want to use Llama.cpp directly. I'm not sure if they're already doing this. Maybe the previous default container run command also works, we could try it out.

Comment (kjjuno) [reply lua/gen/init.lua:219]: I guess we could always restore original functionality and only support context from one chat to another if curl is available. I’ll try to play with that idea in the next few days

Comment (David-Kunz) [reply lua/gen/init.lua:219]: Okay, that's also something for a follow up PR

**Comment by David-Kunz** [comment]
Thanks a ton, @kjjuno, highly appreciated!

**Comment by kjjuno** [comment]
> Thanks a ton, @kjjuno, highly appreciated!

Glad to help! Thanks for taking the addition. It was a really fun project to work on

**Pull request merged**